File: java/plzrun/Boj/src/dp/Boj11055.java
package dp;
import java.util.*;

public class Boj11055 {
	static int[] cash;
	static int[] dp;
	
	public static void main(String[] args) {
		Scanner scanner=new Scanner(System.in);
		
		int number;
		int max=Integer.MIN_VALUE;
		number=scanner.nextInt();
		
		cash=new int[number+1];
		dp=new int[number+1];
		
		for(int i=1; i<=number; i++) {
			cash[i]=scanner.nextInt();
		}
		
		dp[0]=0;
		
		if(number>=1) {
			for(int i=1; i<=number; i++) {
				dp[i]=cash[i];
				for(int j=1; j<i; j++) {
					if(cash[j]<cash[i])
						dp[i]=Math.max(dp[j]+cash[i], dp[i]);
				}
				
			}
			
		}
		for(int i=1; i<=number; i++)
			if(max<dp[i])
				max=dp[i];
		System.out.println(max);
		
		
		scanner.close();
		return;
	}
}

File: c++/plzrun/DpCpp/Boj11055/Boj11055.cpp
#include <iostream>

using namespace std;

int dp[1001];
int input[1001];

int main() {
	int number;
	int max = 0;

	dp[0] = 0;
	input[0] = 0;

	cin >> number;
	for (int i = 1; i <= number; i++) {
		cin >> input[i];
		dp[i] = input[i];
		for (int j = 1; j < i; j++)
			if (input[i] > input[j]&& dp[j] + input[i]>dp[i])
				dp[i] = dp[j] + input[i];
		if (dp[i] > max)
			max = dp[i];
	}

	cout << max;

	return 0;
}